Mathlib/CategoryTheory/Bicategory/Functor.lean

```diff
@@ -0,0 +1,566 @@
+
+/-
+We use this auxiliary definition instead of writing it directly in the definition
+of oplax functors because doing so will cause a timeout.
+-/
+/-- This auxiliary definition states that oplax functors preserve the associators
+modulo some adjustments of domains and codomains of 2-morphisms.
+-/
+@[simp]
+def OplaxFunctor.Map₂AssociatorAux (obj : B → C) (map : ∀ {X Y : B}, (X ⟶ Y) → (obj X ⟶ obj Y))
+    (map₂ : ∀ {a b : B} {f g : a ⟶ b}, (f ⟶ g) → (map f ⟶ map g))
+    (map_comp : ∀ {a b c : B} (f : a ⟶ b) (g : b ⟶ c), map (f ≫ g) ⟶ map f ≫ map g) {a b c d : B}
+    (f : a ⟶ b) (g : b ⟶ c) (h : c ⟶ d) : Prop :=
+  map₂ (α_ f g h).hom ≫ map_comp f (g ≫ h) ≫ map f ◁ map_comp g h =
+    map_comp (f ≫ g) h ≫ map_comp f g ▷ map h ≫ (α_ (map f) (map g) (map h)).hom
+#align category_theory.oplax_functor.map₂_associator_aux CategoryTheory.OplaxFunctor.Map₂AssociatorAux
+
+/-- An oplax functor `F` between bicategories `B` and `C` consists of a function between objects
+`F.obj`, a function between 1-morphisms `F.map`, and a function between 2-morphisms `F.map₂`.
+
+Unlike functors between categories, `F.map` do not need to strictly commute with the composition,
+and do not need to strictly preserve the identity. Instead, there are specified 2-morphisms
+`F.map (𝟙 a) ⟶ 𝟙 (F.obj a)` and `F.map (f ≫ g) ⟶ F.map f ≫ F.map g`.
+
+`F.map₂` strictly commute with compositions and preserve the identity. They also preserve the
+associator, the left unitor, and the right unitor modulo some adjustments of domains and codomains
+of 2-morphisms.
+-/
+structure OplaxFunctor (B : Type u₁) [Bicategory.{w₁, v₁} B] (C : Type u₂)
+  [Bicategory.{w₂, v₂} C] extends PrelaxFunctor B C where
+  map_id (a : B) : map (𝟙 a) ⟶ 𝟙 (obj a)
+  map_comp {a b c : B} (f : a ⟶ b) (g : b ⟶ c) : map (f ≫ g) ⟶ map f ≫ map g
+  mapComp_naturality_left' :
+    ∀ {a b c : B} {f f' : a ⟶ b} (η : f ⟶ f') (g : b ⟶ c),
+      map₂ (η ▷ g) ≫ map_comp f' g = map_comp f g ≫ map₂ η ▷ map g := by
+    aesop
+  mapComp_naturality_right' :
+    ∀ {a b c : B} (f : a ⟶ b) {g g' : b ⟶ c} (η : g ⟶ g'),
+      map₂ (f ◁ η) ≫ map_comp f g' = map_comp f g ≫ map f ◁ map₂ η := by
+    aesop
+  map₂_id' : ∀ {a b : B} (f : a ⟶ b), map₂ (𝟙 f) = 𝟙 (map f) := by aesop
+  map₂_comp' :
+    ∀ {a b : B} {f g h : a ⟶ b} (η : f ⟶ g) (θ : g ⟶ h), map₂ (η ≫ θ) = map₂ η ≫ map₂ θ := by
+    aesop
+  map₂_associator' :
+    ∀ {a b c d : B} (f : a ⟶ b) (g : b ⟶ c) (h : c ⟶ d),
+      OplaxFunctor.Map₂AssociatorAux obj map map₂ map_comp f g h := by
+    aesop
+  map₂_left_unitor' :
+    ∀ {a b : B} (f : a ⟶ b),
+      map₂ (λ_ f).hom = map_comp (𝟙 a) f ≫ map_id a ▷ map f ≫ (λ_ (map f)).hom := by
+    aesop
+  map₂_right_unitor' :
+    ∀ {a b : B} (f : a ⟶ b),
+      map₂ (ρ_ f).hom = map_comp f (𝟙 b) ≫ map f ◁ map_id b ≫ (ρ_ (map f)).hom := by
+    aesop
+#align category_theory.oplax_functor CategoryTheory.OplaxFunctor
+
+namespace OplaxFunctor
+
+restate_axiom mapComp_naturality_left'
```

Comment:
You can get rid of `restate_axiom` and the primes.

Reply:
I don't really understand what's going on here. Do the same comments apply to the 9 other instances of `restate_axiom` too?

Reply:
Yeah. I’ve been yanking them out and it hasn’t but yet.

Reply:
OK I've removed all of them.